File: src/servlets/AdminEditSection.java
import java.io.*;
import java.util.*;

import javax.servlet.*;
import javax.servlet.http.*;

import freemarker.template.*;

import except.*;
import model.ctrl.*;
import model.entity.*;

public class AdminEditSection extends HttpServlet {
  static  String templateDir = "WEB-INF/templates";
  private Configuration  cfg;

  public void init() {
    // Prepare the FreeMarker configuration;
    // - Load templates from the WEB-INF/templates directory of the Web app.
    //
    cfg = new Configuration();
    cfg.setServletContextForTemplateLoading( getServletContext(), "WEB-INF/templates");
  }
  
  public void doGet( HttpServletRequest  req, HttpServletResponse res ) throws ServletException, IOException {
 	Template resultTemplate = null;
    BufferedWriter out = null;
    User u = null;
    String resultTemplateName = "AdminEditSection.ftl",
    		error = null;

    HttpSession session = req.getSession();
	u = (User) session.getAttribute( "user" );
    
    // Setup the data-model
    //
    Map vars = new HashMap();

    // Build the data-model
    //
    if(u == null) resultTemplateName = "Login.ftl";
    
    // Load templates from the WEB-INF/templates directory of the Web app.
    //
    try {
      resultTemplate = cfg.getTemplate( resultTemplateName );
    }catch (IOException e) {
      throw new ServletException("Can't load template in: " + templateDir + ": " + e.toString());
    }

    // Prepare the HTTP response:
    // - Use the charset of template for the output
    // - Use text/html MIME-type
    //
    out = new BufferedWriter(new OutputStreamWriter( res.getOutputStream(), resultTemplate.getEncoding() ));
    
    res.setContentType("text/html; charset=" + resultTemplate.getEncoding());

    // Merge the data-model and the template
    //
    try {
      resultTemplate.process( vars, out );
      out.flush();
    }catch (TemplateException e) {
      throw new ServletException( "Error while processing template", e);
    }
    out.close();
  }

  public void doPost( HttpServletRequest  req, HttpServletResponse res ) throws ServletException, IOException {
  	PrintWriter out = res.getWriter();
    res.setContentType("text/plain; charset=UTF-8");
    
    String dept = req.getParameter( "dept" );
    String course = req.getParameter( "course" );
    String section = req.getParameter( "section" );
    String modify = req.getParameter( "modify" );
    String create = req.getParameter( "create" );
    String delete = req.getParameter( "delete" );
    String faculty = req.getParameter( "faculty" );
    String search = req.getParameter( "search" );
    String days = req.getParameter( "days" );
    String period = req.getParameter( "period" );
    
    //HttpSession session = req.getSession();
	//User u = (User) session.getAttribute( "user" );
    
    StringBuffer content = new StringBuffer();
    
	content.append("{\n");	content.append("  \"results\": {\n");	content.append("    \"bindings\": [\n\n");
    
    if(dept != null){
    	try{
	    	for(Iterator iter = CtrlRegistration.getDepartments(); iter.hasNext();){
	    		Department d = (Department) iter.next();
	    		content.append("      {\n");				content.append("        \"department\": { \"id\": \""+d.getId()+"\", \"name\": \""+d.getName()+"\"  }\n");				content.append("      } ,\n");
	    	}
	    	content.setCharAt(content.length()-2, ' ');
    	}catch(PortsException e){}
    }else if(modify != null && course != null && faculty != null && days != null && period != null){
    	content.append("      {\n");
    	try{
    		int sectionId = Integer.parseInt(modify);
	    	int courseId = Integer.parseInt(course);
	    	int facultyId = Integer.parseInt(faculty);
	    	CtrlAdminEditSection.modifySection(sectionId, courseId, facultyId, days, period);
	    	content.append("        \"success\": \"true\"\n");
	    }catch(PortsException e){content.append("        \"success\": \"false\", \"error\": \""+e.toString()+"\" \n");}
	    content.append("      } ,\n");
    }else if(create != null && course != null && faculty != null && days != null && period != null){
    	content.append("      {\n");
    	try{
	    	int courseId = Integer.parseInt(course);
	    	int facultyId = Integer.parseInt(faculty);
	    	CtrlAdminEditSection.createSection(courseId, facultyId, days, period);
	    	content.append("        \"success\": \"true\"\n");
	    }catch(PortsException e){content.append("        \"success\": \"false\"\n");}
	    content.append("      } ,\n");
    }else if(course != null){
    	try{
	    	for(Iterator iter = CtrlRegistration.getCourses(Integer.parseInt(course)); iter.hasNext();){
	    		Course c = (Course) iter.next();
	    		content.append("      {\n");				content.append("        \"course\": { \"id\": \""+c.getId()+"\", \"dept\": \""+
											c.getDeptId()+"\", \"name\": \""+c.getName()+"\", \"desc\": \""+c.getDesc()+"\" }\n");				content.append("      } ,\n");
	    	}
	    	content.setCharAt(content.length()-2, ' ');
	    }catch(PortsException e){}
    }else if(search != null){
    	try{
	    	for(Iterator iter = CtrlAdminEditSection.getFaculty(search); iter.hasNext();){
	    		Faculty f = (Faculty) iter.next();
	    		content.append("      {\n");				content.append("        \"faculty\": { \"id\": \""+f.getId()+"\", \"name\":\""+
										f.getName()+"\", \"type\": \""+f.getTitle()+"\", \"address\": \""+
										f.getAddress()+"\"  }\n");				content.append("      } ,\n");
	    	}
	    	content.setCharAt(content.length()-2, ' ');
	    }catch(PortsException e){}
    }else if(section != null){
    	try{
	    	for(Iterator iter = CtrlRegistration.getSections(Integer.parseInt(section)); iter.hasNext();){
	    		Section s = (Section) iter.next();
	    		Faculty f = s.getFaculty();
	    		Course c = s.getCourse();
	    		String facName = f.getName();
	    		content.append("      {\n");				content.append("        \"section\": { \"id\": \""+s.getId()+"\", \"course\":\""+
										s.getCourseId()+"\", \"courseName\":\""+c.getName()+"\", \"facId\":\""+f.getId()+"\", \"facName\": \""+facName+"\", \"days\": \""+
										s.getDays()+"\", \"period\": \""+s.getPeriod()+"\"  }\n");				content.append("      } ,\n");
	    	}
	    	content.setCharAt(content.length()-2, ' ');
	    }catch(PortsException e){}
    }else if(delete != null){
    	content.append("      {\n");
    	try{
	    	int sectionId = Integer.parseInt(delete);
	    	CtrlAdminEditSection.deleteSection(sectionId);
	    	content.append("        \"success\": \"true\"\n");
	    }catch(PortsException e){content.append("        \"success\": \"false\"\n");}
	    content.append("      } ,\n");
    }
    content.append("    ]\n");	content.append("  }\n");	content.append("}\n");
    out.println(content.toString());
  }

}

File: web/js/adminEditSection.js
var sections = {};
var courses = {};

var base = "AdminEditSection?";

function getDepartments(){
	$("#status")[0].innerHTML = "";
	var url=base+"dept";
	
	var callback = function(data) {
		var json = parseJSON(data);
		var b = json["results"]["bindings"];
		var html = "<option value='-1'>--</option>";
		for(var i=0; i<b.length; i++){
			var dept = b[i]["department"];
			html += "<option value='"+dept.id+"'>"+dept.name+"</option>\n";
		}
		$("#deptMenu")[0].innerHTML = html;
	};
	
	$.post(url, callback);
}

function getCourses(deptId){
	$("#status")[0].innerHTML = "";
	$("#section")[0].innerHTML = "";
	if(deptId != -1){
		var url=base+"course="+deptId;
		
		var callback = function(data) {
			courses = {};
			var json = parseJSON(data);
			var b = json["results"]["bindings"];
			var html = "<option value='-1'>--</option>";
			for(var i=0; i<b.length; i++){
				var course = b[i]["course"];
				html += "<option value='"+course.id+"'>"+course.name+"</option>\n";
				courses[course.id] = {"dept":course.dept, "name":course.name, "desc":course.desc};
			}
			$("#courseMenu")[0].innerHTML = html;
		};
		
		$.post(url, callback);
	}
}

function getSections(courseId){
	if(courseId != -1){
		var url=base+"section="+courseId;
		
		var callback = function(data) {
			sections = {};
			var json = parseJSON(data);
			var b = json["results"]["bindings"];
			var html = "<option value='-1'>--</option>";
			for(var i=0; i<b.length; i++){
				var section = b[i]["section"];
				html += "<option value='"+section.id+"'>"+section.facName+": "+section.days+"-"+section.period+"</option>\n";
				sections[section.id] = {"course": section.course, "courseName": section.courseName,"facId": section.facId, "facName": section.facName, "days": section.days, "period": section.period};
			}
			$("#sectionMenu")[0].innerHTML = html;
			$("#courseInfo")[0].innerHTML = section.courseName;
		};
		
		$.post(url, callback);
	}
}

function searchFaculty(facName){
	var url=base+"search="+facName;
	
	var callback = function(data) {
		var json = parseJSON(data);
		var b = json["results"]["bindings"];
		
		var html = "<table><tbody><tr><th>Id</th><th>Name</th><th>Title</th><th>Address</th></tr>";
		for(var i=0; i<b.length; i++){
			var faculty = b[i]["faculty"];
			html += "<tr><td>"+faculty.id+"</td><td>"+faculty.name+"</td><td>"+
			faculty.type+"</td><td>"+faculty.address+"</td>"+
			"<td><input type='button' value='Assign' onclick=\"assignFaculty("+faculty.id+", '"+faculty.name+"')\"/></td></tr>\n";
		}
		html +="</table>";
		$("#facultyResults")[0].innerHTML = html;
	};
	$.post(url, callback);
}

function assignFaculty(facId, facName){
	$("#faculty")[0].value = facId;
	$("#facultyInfo")[0].innerHTML = facName;
}

function deleteSection(){
	var section = $("#section")[0];
	var course = $("#course")[0];
	if(section.value != ""){
		var url=base+"delete="+section.value;
		var callback = function(data) {
			var json = parseJSON(data);
			var b = json["results"]["bindings"];
			var success = b[0]["success"]=="true"?"Deletion Successful": "Deletion Unsuccessful";
			$("#status")[0].innerHTML = success;
			var courseMenu = $("#courseMenu")[0];
			getSections(courseMenu.options[courseMenu.selectedIndex].value);
			clearForm();
		};
		$.post(url, callback);
	}
}

function displaySectionInfo(sectionId){
	if(sectionId != -1){
		var section = sections[sectionId];
		
		$("#section")[0].value = sectionId;
		$("#course")[0].value = section.course;
		$("#faculty")[0].value = section.facId;
		$("#days")[0].value = section.days;
		$("#period")[0].value = section.period;
		$("#courseInfo")[0].innerHTML = section.courseName;
		$("#facultyInfo")[0].innerHTML = section.facName;
		$("#sectionInfo")[0].innerHTML = sectionId;
		
		$("#action")[0].value = "Modify Section";
		$("#action")[0].onclick = modifySection;
	}
}

function modifySection(){
	var section = $("#section")[0];
	var course = $("#course")[0];
	var faculty = $("#faculty")[0];
	var days = $("#days")[0];
	var period = $("#period")[0];
	if(section.value != "" && course.value != "" && faculty.value != "" && days.value != "" && period.value != ""){
		var url=base+"modify="+section.value+"&course="+course.value +
						"&faculty="+faculty.value+"&days="+days.value+"&period="+period.value;
		var callback = function(data) {
			var json = parseJSON(data);
			var b = json["results"]["bindings"];
			var success = b[0]["success"]=="true";
			var message = success?"Modification Successful": "Modification Unsuccessful";
			$("#status")[0].innerHTML = message;
			if(success) clearForm();
			var courseMenu = $("#courseMenu")[0];
			getSections(courseMenu.options[courseMenu.selectedIndex].value);
		};
		$.post(url, callback);
	}else{
		alert("Specify all fields to modify section");
	}
}

function createSection(){
	var course = $("#course")[0];
	var faculty = $("#faculty")[0];
	var days = $("#days")[0];
	var period = $("#period")[0];
	if(course.value != "" && faculty.value != "" && days.value != "" && period.value != ""){
		var url=base+"create&course="+course.value+
						"&faculty="+faculty.value+"&days="+days.value+"&period="+period.value;
		var callback = function(data) {
			var json = parseJSON(data);
			var b = json["results"]["bindings"];
			
			var success = b[0]["success"]=="true"
			var message = success?"Creation Successful": "Creation Unsuccessful";
			$("#status")[0].innerHTML = message;
			if(success) clearForm();
			var courseMenu = $("#courseMenu")[0];
			getSections(courseMenu.options[courseMenu.selectedIndex].value);
		};
		$.post(url, callback);
	}else{
		alert("Specify all fields to create section");
	}
}

function clearForm(){
	$("#section")[0].value = "";
	$("#faculty")[0].value = "";
	$("#days")[0].value = "";
	$("#period")[0].value = "";
	//$("#courseInfo")[0].innerHTML = "";
	$("#facultyInfo")[0].innerHTML = "";
	$("#sectionInfo")[0].innerHTML = "";
	
	//var courseMenu = $("#courseMenu")[0];
	//$("#course")[0].value = courseMenu.options[courseMenu.selectedIndex];
	
	$("#action")[0].value = "Create New Section";
	$("#action")[0].onclick = createSection;
}

function parseJSON(json){
	return eval("("+json+")");
}